Presentation: Common rule docs (#3186)
iTwin/itwinjs-core: docs/presentation/RepeatableRelationshipPathSpecification.md

### Jumping through the same relationship multiple times

```ts
[[include:RepeatableRelationshipPathSpecification.SingleStepWithCount.Ruleset]]
```

![Content of the grand-parent element](./media/repeatablerelationshippathspecification-singlestep-with-count.png)

### Jumping through the relationship recursively

```ts
[[include:RepeatableRelationshipPathSpecification.RecursiveSingleStep.Ruleset]]
```

When the root subject is provided as input, content for all its child elements is returned:

![Content of all root subject's child elements](./media/repeatablerelationshippathspecification-recursivesinglestep.png)

### Combining recursive and non-recursive steps

```ts
[[include:RepeatableRelationshipPathSpecification.RecursiveAndNonRecursiveSpecificationsCombination.Ruleset]]
```

When a physical model is provided as input, categories' content of all its elements and their children is returned:

![Categories' content of model elements and their children](./media/repeatablerelationshippathspecification-combinedsteps.png)

### Combining multiple recursive steps

```ts
[[include:RepeatableRelationshipPathSpecification.MultipleRecursiveSpecificationsCombination.Ruleset]]
```

> ![Content of multiple recursive relationship steps](./media/repeatablerelationshippathspecification-combinedrecursivesteps.png)

Reviewer: I'm asking this because this page doesn't explain or I perhaps missed it. Why the result of the last example includes instances of intermediate path steps while all other examples don't?

Author: Because `count: "*"` means that the step is optional (we're making the step 0+ times). Which means we're taking results of the previous step + any instances we find making the step recursively.

Reviewer: Could you please extend `count` attribute section with another example where the relationship path consists of two specifications with `"*"` count?

Author: Updated, please have a look

Reviewer: I see that this pattern gets complicated really quickly, so I don't think we can cover everything with just one example.

Maybe `count` attribute documentation should only say that the special value means an unbounded number of recursive traversals (including zero), and describe the logic behind recursive traversal in another section. 

The examples section could be renamed to "Recursive traversal rules" or similar. The existing examples cover some, but not all cases:

* Fixed number of recursive traversals (covered)
* Unbounded number of recursive traversals with `"*"` count (covered)
* Combination of the above two cases (covered)
* Two unbounded traversals, first of which can be skipped given the right class and relationship structure
* Two unbounded traversals, second of which can be skipped given the right class and relationship structure
* Case where two different instance types are returned (covered)

Author: > Two unbounded traversals, first of which can be skipped given the right class and relationship structure (can this even happen and does it make any sense?)

Yes, this can happen. An example could be `ElementOwnsChildElements + ElemenRefersToElements` where the input is a parent element. The recursive step of `ElementOwnsChildElements` may return nothing and in that case only the given parent element would be used as input to the `ElemenRefersToElements` step.

Problem is that our test iModel is not suitable to reproduce this kind of case, so I can't provide a snippet that can be tested. I updated the description of the `count: "*"` meaning - hopefully that will be sufficient. Please have a look.

Reviewer: It is better than before, thanks

# Repeatable Relationship Path Specification

> TypeScript type: [RepeatableRelationshipPathSpecification]($presentation-common).

> Repeatable relationship path specification is used to define a relationship path to an ECClass, optionally traversing through the same relationship
> multiple times or recursively.

Reviewer: ```suggestion
This specification allows users to specify an indirect relationship between ECInstances. Indirect relationship is expressed as a chain of relationship classes which can optionally be traversed recursively.
```

Author: Can you explain why you suggest this specification is to specify an "**indirect** relationship"?

Reviewer: I posted an alternative suggestion, I would like to have your feedback on it

> Repeatable relationship path specification is used to define a relationship path to an ECClass, optionally traversing through the same relationship
> multiple times or recursively.

Reviewer: ```suggestion
This specification declares a step in a relationship path between a source and target ECInstances. A step can optionally be repeated a number of times to traverse the same relationship recursively. Multiple specifications of this type can be chained together to express complex indirect relationships.
```

| [`count`](#attribute-count)               | No        | `number \| "*"`                  | `1`                           |

### Attribute: `relationship`

This attribute specifies the ECRelationship that should be used to traverse to target class.

### Attribute: `direction`

This attribute specifies the direction in which the [relationship](#attribute-relationship) should be followed:

- `"Forward"` - the relationship is traversed from source to target of the relationship.
- `"Backward"` - the relationship is traversed from target to source of the relationship.

### Attribute: `targetClass`

> **Default value:** Target ECClass of the [relationship](#attribute-relationship) if the [direction](#attribute-direction) is `"Forward"` or
> source ECClass if the [direction](#attribute-direction) is `"Backward"`.

This attribute may be used to specialize the target of the relationship. E.g. when relationship points to a class like `bis.Element`, this
attribute allows specializing it to `bis.PhysicalElement` or some other `bis.Element` subclass.

### Attribute: `count`

> **Default value:** `1`

> This attribute specifies the number of times the relationship should be traversed.
> 
> The special `"*"` value makes the step recursive, which means the relationship is traversed as long as new instances are found by
> using output of the previous step as input for the current step. Outputs of each step traversal are accumulated, combined with outputs
> of the previous step (if any) and passed as input to the next step or used as path output if there are no more steps.

Reviewer: ```suggestion
When a number is specified, the relationship is traversed recursively the specified number of times.

When it is set to a special value `"*"`, the same relationship is traversed recursively unbounded number of times, starting from zero (the relationship is not followed). On each traversal iteration, Presentation rules engine accumulates all indirectly related ECInstances as defined by the remaining relationship path.
```

## Examples

When the [`count` attribute](#attribute-count) is omitted or set to `1`, the specification works similarly to [RelationshipPathSpecification](./RelationshipPathSpecification.md). See its [examples section](./RelationshipPathSpecification.md#examples) for those simpler cases.

> ### Jumping through the same relationship multiple times

Reviewer: ```suggestion
### Jumping through the same relationship recursively fixed number of times
```

```ts
[[include:RepeatableRelationshipPathSpecification.SingleStepWithCount.Ruleset]]
```

![Content of the grand-parent element](./media/repeatablerelationshippathspecification-singlestep-with-count.png)

> ### Jumping through the relationship recursively

Reviewer: ```suggestion
### Jumping through the relationship recursively unbounded number of times
```

```ts
[[include:RepeatableRelationshipPathSpecification.RecursiveSingleStep.Ruleset]]
```

When the root subject is provided as input, content for all its child elements is returned:

![Content of all root subject's child elements](./media/repeatablerelationshippathspecification-recursivesinglestep.png)

### Combining recursive and non-recursive steps

```ts
[[include:RepeatableRelationshipPathSpecification.RecursiveAndNonRecursiveSpecificationsCombination.Ruleset]]
```

When a physical model is provided as input, categories' content of all its elements and their children is returned:

![Categories' content of model elements and their children](./media/repeatablerelationshippathspecification-combinedsteps.png)

> ### Combining multiple recursive steps

Reviewer: ```suggestion
### Combining multiple unbounded recursive steps
```
